Encryption.md

```diff
@@ -0,0 +1,184 @@
+
+## Encryption algorithms
+
+Parquet encryption algorithms are based on the standard AES ciphers for symmetric encryption. 
+AES is supported in Intel and other CPUs with hardware acceleration of 
+crypto operations (“AES-NI”) - that can be leveraged by e.g. Java programs (automatically 
+via HotSpot), or C++ programs (via EVP-* functions in OpenSSL).
+
+Initially, two algorithms are implemented, one based on AES-GCM cipher, and the other on a 
+combination of AES-GCM and AES-CTR ciphers.
+
+AES-GCM is an authenticated encryption. Besides the data confidentiality (encryption), it 
+supports two levels of integrity verification / authentication: of the data (default), and 
+of the data combined with an optional AAD (“additional authenticated data”). The default 
+authentication allows to make sure the data has not been tampered with. An AAD is a free 
+text to be signed, together with the data. The user can, for example, pass the file name 
+with its version (or creation timestamp) as the AAD, to verify the file has not been 
+replaced with an older version.
+
+Sometimes, a hardware acceleration of AES is unavialable (e.g. in Java 8). Then AES crypto 
+operations are implemented in software, and can be somewhat slow, becoming a performance 
+bottleneck in certain workloads. AES-CTR is a regular (not authenticated) cipher.  
+It is faster than AES-GCM, since it doesn’t perform integrity verification and doesn’t 
+calculate the authentication tag. For applications running without AES acceleration and 
+willing to compromise on content verification, AES-CTR can provide a boost in Parquet 
+writing/reading throughput. The second Parquet algorithm encrypts the data content (pages) 
+with AES-CTR, and the metadata (Thrift structures) with AES-GCM. This allows to encrypt/decrypt 
+the bulk of the data faster, while still verifying the metadata integrity and making sure 
+the file had not been replaced with a wrong version. However, tampering with the page data 
+might go unnoticed. 
+
+
+### AES_GCM_V1
+All modules are encrypted with the AES-GCM cipher. The authentication tags (16 bytes) are 
+written after each ciphertext. The IVs (12 bytes) are either written before each ciphertext, 
+or split into two parts: a fixed part (n bytes) written in the FileCryptoMetaData structure 
+(iv_prefix field), and a variable part (12-n bytes: counter, random, etc) written before 
+each ciphertext.
+
+### AES_GCM_CTR_V1
+Thrift modules are encrypted with the AES-GCM cipher, as described before. The pages are 
+encrypted with AES-CTR, where the IVs (16 bytes) are either written before each ciphertext, 
+or split into two parts: a fixed part (n bytes) written in the FileCryptoMetaData structure 
+(iv_prefix field), and a variable part (16-n bytes: counter, random, etc) written before 
+each ciphertext.
+
+
+
+## File Format
+
+The encrypted Parquet files have a different magic string (“PARE”), and an extension 
+(“parquet.encrypted”).
```

Comment:
If we allow the old reader to read the unencrypted columns, are the extension and magic string needed to be changed?